### PR TITLE
Do grid detection in Artemis

### LINKED ISSUE #373: OAV: get gridscan grid from waveforms
Artemis should be able to get the grid to do the 3D xray centre based on the OAV image.

## Acceptance Criteria
* The grid size etc. is determined from the OAV image (for 0 and 90)
* Snapshots with grids are taken through Artemis
* This data is passed on correctly to existing code, which performs the grids
* GDA is able to trigger this, as well as just the xray centre with provided parameters
* There are sufficient system/unit tests around this

### LINKED ISSUE #652: Run Grid Detection and Grid in Artemis
Following https://github.com/DiamondLightSource/python-artemis/issues/554 we have a plan that we can trigger from GDA in exactly the same way we trigger the `fast_grid_scan` and it will calculate the grids in Artemis as well as in GDA then use the GDA values. We should move this to a full scan happening in Artemis e.g:
* The plan must only accept the values it needs
* It will do the grid detection, the full grid and all the work in between (https://github.com/DiamondLightSource/python-artemis/issues/567)
* It should be easy to turn this on/off in GDA

### LINKED ISSUE #658: full grid scan testing notes
- aperture z check for safe move has no tolerance (??!!)
- running once works, twice does not
- pin tip centring dosn't happen at start of full grid scan (do we want it to?)
- sometimes num_images is negative (??!!) https://graylog2.diamond.ac.uk/messages/logs_8246/27ff3cc1-f308-11ed-9360-1866dafadc6c -> sometimes num_steps is negative - was default values for min and max grid size the wrong way round - we are hitting edge of image somtimes
- centre often wrong and snapshots often not displayed

### PR DESCRIPTION
**Pull request by DominicOram**
Fixes #652, fixes #670, fixes #373, fixes #658

This has been tested on the beamline and works sufficiently, it could use some tidying though. See #671, #672, #673, #674. Please review bearing these in mind

### To test:
1. Tests pass

### CODE REVIEW

File: src/artemis/experiment_plans/full_grid_scan.py [hunk 1]
```diff
@@ -38,7 +42,7 @@ def create_devices():
     i03.aperture_scatterguard().wait_for_connection()
 
 
-def wait_for_det_to_finish_moving(detector: DetectorMotion, timeout=2):
+def wait_for_det_to_finish_moving(detector: DetectorMotion, timeout=120):
```
Comment (dperl-dls) [inline src/artemis/experiment_plans/full_grid_scan.py:45]: does it really need to be this long?

**Review by dperl-dls — APPROVED** [review]
Looks good, tests pass, thanks!

File: src/artemis/experiment_plans/full_grid_scan.py (hunk 1 above)
Comment (DominicOram) [reply src/artemis/experiment_plans/full_grid_scan.py:45]: Yep, the detector motion is slow, @neil-i03 said that 2mins is about right

Comment (DominicOram) [reply src/artemis/experiment_plans/full_grid_scan.py:45]: No, it's just how long until it will timeout. If the motion finishes earlier it will move on

**Pull request merged**